File: python/rpcClient/infiApiHttp.py
import hashlib
import time
import requests
import hmac

class InfiApiHttpClient:
    def __init__(self, appId, signKey, infiWbsPath):
        self.appId = appId
        self.signKey = signKey
        self.infiWbsPath = infiWbsPath

    def calculate_wsb_params(self, params):
        # 在业务参数的基础上补充签名参数
        params['appId'] = self.appId
        params['expire'] = str((int(time.time()) + 60)*1000)  # 60秒内地址有效

        # 将参数按字典序排序
        sorted_params = sorted(params.items())

        # 连接参数字符串
        content = '&'.join(f'{key}={value}' for key, value in sorted_params)
        print(content)
        # 计算签名
        hmac_obj = hmac.new(self.signKey.encode(), content.encode(), hashlib.sha1)
        signature = hmac_obj.hexdigest().upper()
        params['signature'] = signature
        print(signature)
        return content + '&signature=' + signature

    def calculate_balance_params(self, params):
        # 在业务参数的基础上补充签名参数
        params['appId'] = self.appId
        params['validBegin'] = str((int(time.time())))  # 60秒内地址有效
        params['validTime'] = "120"  # 120秒内地址有效
        # 将参数按字典序排序
        sorted_params = sorted(params.items())

        # 连接参数字符串
        content = '&'.join(f'{key}={value}' for key, value in sorted_params)
        print(content)
        # 计算签名
        hmac_obj = hmac.new(self.signKey.encode(), content.encode(), hashlib.sha1)
        signature = hmac_obj.hexdigest().upper()
        params['signature'] = signature
        print(signature)
        return content + '&signature=' + signature

    def create_whiteboard(self, query, params):
        queryParams = self.calculate_wsb_params(query)
        url = f'{self.infiWbsPath}/u3wbs/wbs/nc/createBoard?'+queryParams
        print(url)
        try:
            response = requests.post(url, json=params)
            return response.json()
        except requests.exceptions.RequestException as error:
            print('Error creating whiteboard:', error)
            return None

File: node/rpcClient/infiApiHttp.js
const crypto = require('crypto');
const axios = require('axios');
class InfiApiHttpClient {
    constructor(appId,signKey,infiWbsPath) {
        this.appId = appId;
        this.signKey = signKey;
        this.infiWbsPath = infiWbsPath;
    }
    // 计算api请求的签名
    calculateWbsParams(params) {
        // 在业务参数的基础上补充签名参数
        params.appId = this.appId;
        params.expire = (Date.now() + 60 * 1000).toString(); // 60秒内地址有效

        // 将参数转换为字符串
        const businessParam = {};
        for (const [key, value] of Object.entries(params)) {
            businessParam[key] = String(value);
        }

        // 排序参数
        const keys = Object.keys(businessParam).sort();

        const res = [];
        for (const key of keys) {
            if (businessParam.hasOwnProperty(key)) {
                res.push(`${key}=${businessParam[key]}`);
            }
        }

        // 连接参数字符串
        let content = '';
        if (res.length > 0) {
            content = res.join('&');
        }

        console.log(`content: ${content}`);

        // 计算签名
        const hmac = crypto.createHmac('sha1', this.signKey);
        hmac.update(content);
        const signature = hmac.digest('hex').toUpperCase();

        params.signature = signature;
        console.log(`signature: ${signature}`);
        return `${content}&signature=${signature}`;
    }
    // 计算白板连接的签名
    calculateBalanceParams(params) {
        // 在业务参数的基础上补充签名参数
        params.appId = this.appId;
        params.validBegin = Math.floor(Date.now() / 1000).toString();
        params.validTime = '120'; // 120秒内地址有效

        // 将参数转换为字符串
        const businessParam = {};
        for (const [key, value] of Object.entries(params)) {
            businessParam[key] = String(value);
        }

        // 排序参数
        const keys = Object.keys(businessParam).sort();

        const res = [];
        for (const key of keys) {
            if (businessParam.hasOwnProperty(key)) {
                res.push(`${key}=${businessParam[key]}`);
            }
        }

        // 连接参数字符串
        let content = '';
        if (res.length > 0) {
            content = res.join('&');
        }

        console.log(`content: ${content}`);

        // 计算签名
        const hmac = crypto.createHmac('sha1', this.signKey);
        hmac.update(content);
        const signature = hmac.digest('hex').toUpperCase();

        params.signature = signature;
        return `${content}&signature=${signature}`;
    }
    async createWhiteBoard(query, params) {
        const queryParams = this.calculateWbsParams(query);
        console.log(queryParams);
        const url = this.infiWbsPath + '/u3wbs/wbs/nc/createBoard?'+queryParams;

        try {
            const response = await axios.post(url, params);
            return response.data;
        } catch (error) {
            console.error('Error creating whiteboard:', error);
            return null;
        }
    }

}

module.exports = InfiApiHttpClient;
